src/main/scala/workflow/AutoCacheRule.scala

```diff
@@ -0,0 +1,355 @@
+  }
+
+  def cacheMem(caches: Set[Int], profiles: Map[Int, Profile]): Long = {
+    // Must do a toSeq here otherwise the map may merge equal memories
+    caches.toSeq.map(i => profiles.getOrElse(i, Profile(0, 0)).mem).sum
+  }
+
+  /**
+   * Returns true iff there is still an uncached node whose output is used > once, that would fit in memory
+   * if cached
+   */
+  def stillRoom(caches: Set[Int], runs: Map[Int, Int], profiles: Map[Int, Profile], spaceLeft: Long): Boolean = {
+    runs.exists { i =>
+      (i._2 > 1) &&
+        (!caches.contains(i._1)) &&
+        (profiles.getOrElse(i._1, Profile(0, 0)).mem < spaceLeft)
+    }
+  }
+
+  def selectNext(
+    pipe: Seq[Instruction],
+    profiles: Map[Int, Profile],
+    cached: Set[Int],
+    runs: Map[Int, Int],
+    spaceLeft: Long
+  ): Int = {
+    //Get the uncached node which fits that maximizes savings in runtime.
+    pipe.indices.filter(i =>
+      !cached(i) &&
+        profiles.getOrElse(i, Profile(0, 0)).mem < spaceLeft &&
+        runs(i) > 1)
+      .minBy(i => estimateCachedRunTime(pipe, cached + i, profiles))
+  }
+
+  def greedyCache(
+    instructions: Seq[Instruction],
+    profiles: Map[Int, Profile],
+    maxMem: Long
+  ): Seq[Instruction] = {
+    val nodeWeights = getNodeWeights(instructions)
+
+    var cached = instructions.indices.filter { instructions(_) match {
+      case _: EstimatorNode => true
+      case _: TransformerNode =>  true
+      case _: EstimatorFitNode => true
+      case _: SourceNode => false
+      case _: TransformerApplyNode => false
+    }}.toSet
+
+    var usedMem = cacheMem(cached, profiles)
+    var runs = getRuns(instructions, cached, nodeWeights)
+    while (usedMem < maxMem && stillRoom(cached, runs, profiles, maxMem - usedMem)) {
+      cached = cached + selectNext(instructions, profiles, cached, runs, maxMem - usedMem)
+      runs = getRuns(instructions, cached, nodeWeights)
+      usedMem = cacheMem(cached, profiles)
+    }
+
+    makeCachedPipeline(instructions, cached -- WorkflowUtils.getChildren(Pipeline.SOURCE, instructions))
+  }
+
+  override def apply[A, B](plan: Pipeline[A, B]): Pipeline[A, B] = {
+    val instructions = WorkflowUtils.pipelineToInstructions(plan)
+
+    WorkflowUtils.instructionsToPipeline(cachingMode match {
+      case NaiveCache => naiveCache(instructions)
+      case GreedyCache(maxMem, profileScales, numProfileTrials) => {
+        val profiles = profileInstructions(instructions, profileScales, numProfileTrials)
+        greedyCache(instructions, profiles, maxMem)
+      }
+    })
+  }
+}
+
+object AutoCacheRule {
+  sealed trait CachingStrategy
+  /**
+   * NaiveCache is the optimal caching strategy assuming infinite memory and zero caching overhead
+   * (cache at every node with > 1 direct output dependency edge during training)
+   */
+  case object NaiveCache extends CachingStrategy
+  case class GreedyCache(
+    maxMem: Long,
+    profileScales: Seq[Long] = Seq(500, 1000),
```

Comment:
I honestly don't have any sense of what the right thing to do would be in each of those situations. What's the minimal number of data points needed for each to provide reasonable / useful profiling info?